How to start vLLM on Beverin (using Slurm, Container Engine and Netstack) (#468)
eth-cscs/cscs-docs: docs/tutorials/ml/vllm-amdgpus.md

[](){#software-ml-vllm-amdgpus}

# vLLM on AMD GPUs Tutorial

> This tutorial will guide you through the steps required to setup a vLLM container on AMD GPUs to serve an LLM model.

Reviewer: ```suggestion
This tutorial will guide you through the steps required to setup a vLLM container on AMD GPUs to serve a LLM.
```

In this specific tutorial we are going to show how:

- setup a vLLM container

> - to run an LLM model

Reviewer: ```suggestion
- to run a LLM
```

- with Slurm
- on AMD GPUs
- in a single or multi-node setup
- and test it

!!! warning "What this guide is NOT"

>     This guide is not about how to configure correctly/at best vLLM.

Reviewer: Worth adding a link to vLLM recipes? https://recipes.vllm.ai/

Author: I was thinking more about the setup on the cluster (e.g. fabric configuration, ... ) rather than the actual model setup. I might give a try to be more specific, and I might also cite vLLM recipes in addition to be more specific about cluster performance and setup.

Author: What about something like this?

```suggestion
    This guide is not about how to configure correctly/at best vLLM on Alps clusters.
    For more details about vLLM parameters and best setup for a specific model have a look at <https://recipes.vllm.ai/>, while for what concerns the Alps cluster setup, you'll have to explore documentation for the specific cluster, experiment and benchamrk.
```

Reviewer: Looks great, thanks!

### Prerequisites

> This tutorial assumes you are able to access the cluster via SSH. To set up access to CSCS systems, follow the guide [here][ref-ssh], and read through the documentation about the [ML Platform][ref-platform-mlp].

Reviewer: ```suggestion
This tutorial assumes you are able to access the cluster via SSH. To set up access to CSCS systems, follow the guide [here][ref-ssh], and read through the documentation about the [ML platform][ref-platform-mlp].
```

In particular, this tutorial aims at Beverin cluster, but it should be possible to adapt it with minor changes to other clusters as well.

### Get vLLM image

In this tutorial we are going to use vLLM official image for ROCm systems. From early 2026 the AMD's Docker images [`rocm/vllm`](https://hub.docker.com/r/rocm/vllm) (and others) have been deprecated (see [here](https://docs.vllm.ai/en/v0.19.1/getting_started/installation/gpu/#use-amds-docker-images-deprecated)) in favour of [vllm/vllm-openai-rocm](https://hub.docker.com/r/vllm/vllm-openai-rocm).

Beforehand, if you have not done already, let's create a directory to keep track of all images used with the CE. Since container images are large files and the filesystem is a shared resource, we need to apply [best practices for LUSTRE][ref-guides-storage-lustre] so they are properly distributed across storage nodes.

```console title="Container image directory with recommended LUSTRE settings"
mkdir -p $SCRATCH/ce-images
lfs setstripe -E 4M -c 1 -E 64M -c 4 -E -1 -c -1 -S 4M $SCRATCH/ce-images
```

> Now we can pull the docker image in our container images directory just created. Following command allows to import a docker image as squashfs archive that can be used with [Container Engine][ref-container-engine].

Reviewer: ```suggestion
Now we can pull the docker image in our container images directory just created. The following command allows to import a docker image as squashfs archive that can be used with [Container Engine][ref-container-engine].
```
?

```console
enroot import \
    -o $SCRATCH/ce-images/vllm-opeani-rocm.sqfs \
    docker://vllm/vllm-openai-rocm:latest
```

### Setup EDF

> Following step is creation of the Environment Definition File ([EDF][ref-ce-edf-reference]) where details on how to start the container are specified.

Reviewer: ```suggestion
The following step is the creation of an Environment Definition File ([EDF][ref-ce-edf-reference]) where details on how to start the container are specified.
```
? Not a native speaker as you know, but I think this sounds better?

> In particular, this tutorial makes use of [netstack][ref-ce-netstack-source] and related hooks for binding compatible network libraries (mainly `libfabric`, `CXI` and `aws-ofi-nccl`) inside the container.

Reviewer: Totally unrelated, but I just double checked and there is a `aws-ofi-rccl` package in Spack, while `aws-ofi-nccl` supports both `+cuda` and `+rocm`. Should the former be deprecated?

Reviewer 2: I don't know about deprecation specifically (where?) but yes, `aws-ofi-nccl` should be used for both cuda and rocm nowadays.

Reviewer: > I don't know about deprecation specifically (where?)

I was thinking about Spack upstream. As I said, this comment is totally un-related to the PR, probably I should have moved it to Slack. But the `aws-ofi-rccl` repository itself is deprecated, so it would be good to have it deprecated in Spack too (if package-level deprecations are possible).

Save this as a TOML file named `env-vllm.toml` so that you can use it later.

```toml
image = "/capstor/scratch/cscs/<username>/ce-images/vllm-openai-rocm.sqfs"
writable = true
entrypoint = false

[annotations]
com.hooks.netstack.source = "artifact"
com.hooks.netstack.version = "26.07.1"
com.hooks.netstack.name = "gpu:rocm7,cxi:13.1.0,ofi:2.6.0,aws:1.20.0"

com.hooks.cxi.enabled = "true"
com.hooks.aws_ofi_nccl.enabled = "true"

[env]
# Note: this solves an hang happening at NCCL initialization time
HWLOC_COMPONENTS="-gl"

FI_PROVIDER="cxi"
FI_CXI_RX_MATCH_MODE="software"

> FI_MR_CACHE_MONITOR="disabled"

Reviewer 2: Are these required to avoid hangs as well? We should also update the NCCL/RCCL pages and/or update the CE hooks to use good default values (I don't know if these are required specifically for vLLM only, or if they're useful in general).

Author: It's a good question, that I already tried to figure out before opening the PR...but without clear results 😅

The only one I got consistent results for is just `HWLOC_COMPONENTS`, the others I got mixed results, and they might be superfluous (at least from a functionality point of view, not sure about performance, but it's not the point of this doc page).

```

### Launch a Single Node instance

> At this point everything is ready, CE image needs just to be launched and vLLM started.

Reviewer: ```suggestion
At this point everything is ready, the CE image needs just to be launched and vLLM started.
```

The simplest run possible is a single GPU instance

```console
srun --environment ./env-vllm.toml -pmi300 \
    vllm serve Qwen/Qwen2.5-1.5B-Instruct
```

or using multiple GPUs from the same node

```console

> srun --environment ./env-vllm.toml -pmi300 --gpus-per-task 4\

Reviewer 2: ```suggestion
srun --environment ./env-vllm.toml -pmi300 --gpus-per-task 4 \
```

    vllm serve Qwen/Qwen2.5-1.5B-Instruct --tensor-parallel-size 4
```

### Use the instance

Once vLLM instance is serving, i.e. master node prints out on which address and port is listening, it provides the standard endpoint interface.

> It can be queried for LLM models provided with

Reviewer: ```suggestion
It can be queried for served LLMs with
```
